File: rust/problem0070/src/main.rs
use std::env;

fn perfect_number(n: usize, command: Command) -> usize {
    if n <= 0 || n > 59271 {
        panic!("valid range for n is [1, 59271] got {}", n)
    }
    let mut i = 0;
    let mut current = 19;
    let mut next_power = 100;

    let handle = match command {
        Command::Graph => |i: usize, v: usize| println!("{},{}", i, v),
        Command::None => |_i: usize, _v: usize| (),
    };

    loop {
        if is_perfect(current) {
            handle(i, current);
            i += 1
        }
        if i == n {
            return current;
        }
        if current == next_power {
            eprintln!("debug: passed {}", next_power);
            next_power = next_power * 10;
        }
        current += 1
    }
}

fn is_perfect(n: usize) -> bool {
    let mut current = n;
    let mut sum = 0;
    loop {
        let remaining_places = current / 10;
        if remaining_places == 0 {
            sum += current;
            return sum == 10;
        }
        sum += current - remaining_places * 10;
        current = remaining_places;
    }
}

#[derive(Debug)]
enum Command {
    Graph,
    None,
}

fn main() {
    let args: Vec<String> = env::args().skip(1).collect();

    eprintln!("args {:?}", args);

    let n = match args[0].parse::<usize>() {
        Ok(n) => n,
        Err(err) => panic!(err),
    };

    let command = match args
        .get(1)
        .map(|x| x.to_owned())
        .map(|x| x.to_ascii_lowercase())
        .unwrap_or("none".to_owned())
        .as_ref()
    {
        "graph" => Command::Graph,
        "none" => Command::None,
        _ => panic!("expected `graph` or `none`"),
    };

    eprintln!("command {:?}", command);

    eprintln!("{}", perfect_number(n, command));
}

File: scala/problem0070/src/main/scala/Main.scala
import scala.util.Try
import scala.util.Success
import scala.util.Failure
import scala.util.control._

class NumberOutOfBounds(val n: Int) extends Exception {
  override def getMessage(): String =
    s"Number must be between 0 and 59271, got ${n}"
}
class MissingCLIArgument extends Exception {
  override def getMessage(): String = s"Expected at least 1 argument"
}

object Main extends App {
  val MAX_SEQUENCE_VALUE = 59271
  def eprintln(x: Any) = Console.err.println(x)

  tryIndex(args, 0)
    .flatMap(tryStrToInt)
    .flatMap(validateRange)
    .map(getPerfectNumber) match {
    case Failure(exception) =>
      eprintln(
        s"[${Console.RED}ERROR${Console.RESET}]: ${exception.getMessage()}"
      )
    case Success(value) => println(value)
  }

  //---

  def tryIndex[T](itr: Array[T], idx: Int): Try[T] =
    try { Success(itr(idx)) }
    catch { case e: Exception => Failure(new MissingCLIArgument) }

  def tryStrToInt(str: String): Try[Int] = Try { str.toInt }

  def validateRange(n: Int): Try[Int] =
    n match {
      case n if n <= 0 || n > MAX_SEQUENCE_VALUE =>
        Failure(new NumberOutOfBounds(n))
      case _ => Success(n)
    }

  def getPerfectNumber(sequenceId: Int): Int = {
    var i = 0;
    var current = 19;
    var nextPower = 100;

    while (true) {
      if (isPerfect(current)) {
        i += 1
      }
      if (current == nextPower) {
        eprintln(
          s"[${Console.MAGENTA}DEBUG${Console.RESET}]: passed ${nextPower}"
        );
        nextPower *= 10;
      }
      if (i == sequenceId) {
        return current
      }
      current += 1
    }
    current
  }

  def isPerfect(n: Int): Boolean = {
    var current = n;
    var sum = 0;
    while (true) {
      var remaining_places = current / 10
      if (remaining_places == 0) {
        sum += current
        return sum == 10
      }
      sum += current - remaining_places * 10
      current = remaining_places
    }
    return false
  }

}
